modules/runner/src/testFixtures/java/org/apache/ignite/internal/BaseIgniteRestartTest.java

```diff
@@ -139,21 +146,93 @@ void setUp(TestInfo testInfo) {
     public void afterEachTest() throws Exception {
         var closeables = new ArrayList<AutoCloseable>();
 
-        for (IgniteServer node : IGNITE_SERVERS) {
-            if (node != null) {
-                closeables.add(node::shutdown);
+        List<String> serverNames = IGNITE_SERVERS.stream()
+                .filter(Objects::nonNull)
+                .map(IgniteServer::name)
+                .collect(toList());
+
+        List<String> partialNodeNames = this.partialNodes.stream()
+                .filter(Objects::nonNull)
+                .map(PartialNode::name)
+                .collect(toList());
+
+        log.info("Shutting the cluster down [serverNodes={}, partialNodes={}]", serverNames, partialNodeNames);
+
+        Set<String> cmgMsNodesNames = new HashSet<>();
+
+        Set<String> cmgMsPartialNodesNames = new HashSet<>();
+
+        if (!partialNodes.isEmpty()) {
+            PartialNode anyPartialNode = partialNodes.get(0);
+
+            ClusterManagementGroupManager component = findComponent(
+                    anyPartialNode.startedComponents(),
+                    ClusterManagementGroupManager.class
+            );
+
+            cmgMsPartialNodesNames = msCmgNodes(component);
+
+            for (PartialNode partialNode : partialNodes) {
+                if (!cmgMsPartialNodesNames.contains(partialNode.name())) {
+                    closeables.add(partialNode::stop);
+                }
             }
         }
 
+        if (!IGNITE_SERVERS.isEmpty()) {
+            IgniteServer anyNode = IGNITE_SERVERS.get(0);
+
+            IgniteImpl ignite = unwrapIgniteImpl(anyNode.api());
+
+            cmgMsNodesNames = msCmgNodes(ignite.clusterManagementGroupManager());
+
+            for (IgniteServer node : IGNITE_SERVERS) {
+                if (node != null) {
+                    if (!cmgMsNodesNames.contains(node.name())) {
+                        closeables.add(node::shutdown);
+                    }
+                }
+            }
+        }
+
+        // Add CMG/MS nodes at the end of the list to ensure that they are stopped last.
         if (!partialNodes.isEmpty()) {
             for (PartialNode partialNode : partialNodes) {
```

Comment:
I'd rather have one iteration per partialNodes and one per common nodes within which I'd prepare both cmgMG list and !cmgMg list. When iteration is finished I'd populate closables with cmgMG and !cmgMg in proper order. Meaning that there will be 2 iterations instead of 4.

Please pay attention to may next comment

> It's not guaranteed that closeAll will stop the nodes in proper order, thus, I guess you will need to do it explicitly, meaning that you will need two closable lists.

Reply:
rewritten